registration/manager.go

```diff
@@ -0,0 +1,129 @@
+package registration
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/consul-terraform-sync/client"
+	"github.com/hashicorp/consul-terraform-sync/config"
+	"github.com/hashicorp/consul-terraform-sync/logging"
+	consulapi "github.com/hashicorp/consul/api"
+)
+
+const (
+	// Service defaults
+	defaultServiceName = "Consul-Terraform-Sync"
+	defaultNamespace   = ""
+
+	// Check defaults
+	defaultCheckName                      = "CTS Health Status"
+	defaultCheckNotes                     = "Check created by Consul-Terraform-Sync"
+	defaultDeregisterCriticalServiceAfter = "30m"
+	defaultCheckStatus                    = consulapi.HealthCritical
+
+	// HTTP-specific check defaults
+	defaultEndpoint      = "/v1/status" // TODO: temporary until /v1/health is implemented
+	defaultMethod        = "GET"
+	defaultInterval      = "10s"
+	defaultTimeout       = "2s"
+	defaultTLSSkipVerify = true
+
+	logSystemName = "registration"
+)
+
+var defaultServiceTags = []string{"cts"}
+
+// RegistrationManager handles the registration of a service and the health checks for that service
+// to Consul.
+type RegistrationManager struct {
+	client  client.ConsulClientInterface
+	service *service
+
+	logger logging.Logger
+}
+
+type RegistrationManagerConfig struct {
+	ID               string
+	Port             int
+	TLSEnabled       bool
+	SelfRegistration *config.SelfRegistrationConfig
+}
+
+type service struct {
+	name      string
+	id        string
+	tags      []string
+	port      int
+	namespace string
+
+	checks []*consulapi.AgentServiceCheck
+}
+
+func NewRegistrationManager(conf *RegistrationManagerConfig, client client.ConsulClientInterface) *RegistrationManager {
+	logger := logging.Global().Named(logSystemName)
+
+	name := defaultServiceName
+
+	ns := defaultNamespace
+	if conf.SelfRegistration.Namespace != nil {
+		ns = *conf.SelfRegistration.Namespace
+	}
+
+	var checks []*consulapi.AgentServiceCheck
+	checks = append(checks, defaultHTTPCheck(conf))
+
+	return &RegistrationManager{
+		client: client,
+		logger: logger,
+		service: &service{
+			name:      name,
+			id:        conf.ID,
+			tags:      defaultServiceTags,
+			port:      conf.Port,
+			namespace: ns,
+			checks:    checks,
+		},
+	}
+}
+
+func (m *RegistrationManager) RegisterService(ctx context.Context) error {
+	s := m.service
+	r := &consulapi.AgentServiceRegistration{
+		ID:        s.id,
+		Name:      s.name,
+		Tags:      s.tags,
+		Port:      s.port,
+		Checks:    s.checks,
+		Namespace: s.namespace,
+	}
+
+	m.logger.Debug("registering service with Consul", "name", s.name, "id", s.id)
```

Comment:
_Super_ minor nit: I was curious on your thoughts on updating the language from "registering service" to "registering CTS" (or "registering self"?!)  in some places - e.g. the method name "RegisterService", this log, and the error below. "registering service" initially gives me the impression that CTS is registering other (non-CTS) services.

This is definitely a personal thing - I think it's easily clarified since service name and id are included. Feel free to take or leave!

Reply:
Yeah, I think you're right that this would be better to have CTS/self specific wording! 

I had this more general service language because I realized that this manager could technically be used for any kind of service. I didn't really think we'd ever need to use it to register a service other than CTS, but it just worked out like that. It doesn't make sense to keep it vague for a use case that will likely never exist, though, when having CTS specific logs will be a better experience for the user.

Reply:
Going to rename basically everything to have `SelfRegistration` so that it also better matches the `self_registration` config! I also have realized that I've already baked in some self registration specific stuff (like passing a `config.SelfRegistrationConfig` and having the HTTP default check using CTS specific attributes), so it extra doesn't make sense to keep it as just `Registration`. 🙂